gettsim/taxes/soli_st.py

```diff
@@ -39,16 +42,113 @@ def soli_st_tu(
-            rates=soli_st_params["soli_st"]["rates"],
-            intercepts_at_lower_thresholds=soli_st_params["soli_st"][
-                "intercepts_at_lower_thresholds"
-            ],
-        )
+        anz_erwachsene_tu * soli_tarif(st_per_individual, soli_st_params)
         + soli_st_params["soli_st"]["rates"][0, -1] * abgelt_st_tu
     )
 
     return out
+
+
+def lohn_st_soli(
+    lohn_st_kinderfreibetrag: FloatSeries, soli_st_params: dict
+) -> FloatSeries:
+    """
+    Calculates the Solidarity Surcharge as a top-up on Lohnsteuer
+
+    Parameters
+    ----------
+    lohn_st_kinderfreibetrag
+    soli_st_params
+
+    Returns
+    -------
+
+    """
+
+    return soli_tarif(lohn_st_kinderfreibetrag, soli_st_params)
+
+
+def lohn_st_zve_kifb(
+    lohn_st_zve: FloatSeries, kinderfreibetrag_lohn_st: FloatSeries
+) -> FloatSeries:
+    """ Calculates tax base for Soli Lohnsteuer
+    by subtracting child allowance from regular lohnsteuer taxable income
+    """
+    return np.maximum(lohn_st_zve - kinderfreibetrag_lohn_st, 0)
+
+
+def kinderfreibetrag_lohn_st(
+    tu_id: IntSeries,
+    steuerklasse: IntSeries,
+    anz_kindergeld_kinder_tu: FloatSeries,
+    eink_st_abzuege_params,
+) -> FloatSeries:
+    """ Calculates Child Allowance for Lohnsteuer-Soli
+
+    For the purpose of Soli on Lohnsteuer,
+    the child allowance not only depends on the number of children,
+    but also on the steuerklasse
+    """
+
+    kinderfreibetrag_basis = (
+        eink_st_abzuege_params["kinderfreibetrag"]["sächl_existenzmin"]
+        + eink_st_abzuege_params["kinderfreibetrag"]["beitr_erz_ausb"]
+    )
+
+    # For certain tax brackets, twice the child allowance can be deducted
+    out = (
+        (kinderfreibetrag_basis * 2 * steuerklasse.isin([1, 2, 3]))
+        + (kinderfreibetrag_basis * steuerklasse == 4)
+    ) * tu_id.replace(anz_kindergeld_kinder_tu)
+    return out
+
+
+def lohn_st_kinderfreibetrag(
+    lohn_st_zve_kifb: FloatSeries, steuerklasse: IntSeries, eink_st_params: dict
+) -> FloatSeries:
+    """ Calculate Lohnsteuer just as lohn_st function,
+    but with a different tax base, i.e. including child allowance
+    """
+    lohnsteuer_basistarif = st_tarif(lohn_st_zve_kifb, eink_st_params)
+    lohnsteuer_splittingtarif = 2 * st_tarif(lohn_st_zve_kifb / 2, eink_st_params)
+    lohnsteuer_klasse5_6 = np.maximum(
+        2
+        * (
+            st_tarif(lohn_st_zve_kifb * 1.25, eink_st_params)
+            - st_tarif(lohn_st_zve_kifb * 0.75, eink_st_params)
+        ),
+        lohn_st_zve_kifb * eink_st_params["eink_st_tarif"]["rates"][0][1],
+    )
+
+    out = (
+        (lohnsteuer_splittingtarif * (steuerklasse == 3))
+        + (lohnsteuer_basistarif * (steuerklasse.isin([1, 2, 4])))
+        + (lohnsteuer_klasse5_6 * (steuerklasse.isin([5, 6])))
+    )
+    return out
+
+
+def soli_tarif(st_per_individual: FloatSeries, soli_st_params: dict) -> FloatSeries:
```

Comment:
The function should start with an underscore as it is not part of the DAG, but only called within other functions.

Reply:
See #333 -- not sure we'll even want it in the end.